### PR TITLE
[firebase_auth] Ability to choose between sign-in or link on auto verification of phone number on Android

### PR DESCRIPTION
**Pull request by cqhchan**
- Fixed Android Link with credentials bug.
- Added in ability to choose between sign-in or link on auto verification of phone number

### CODE REVIEW

File: packages/firebase_auth/lib/src/firebase_auth.dart [hunk 1]
```diff
@@ -241,15 +241,20 @@ class FirebaseAuth {
...
+
+  Future<void> verifyPhoneNumber(
+      {@required String phoneNumber,
+      @required Duration timeout,
+      int forceResendingToken,
+      @required PhoneVerificationCompleted verificationCompleted,
+      @required PhoneVerificationFailed verificationFailed,
+      @required PhoneCodeSent codeSent,
+      @required PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout,
+      bool linkCredentials = false}) async {
```
Comment (ened) [inline packages/firebase_auth/lib/src/firebase_auth.dart]: Could you please add back the comma (,) at the end of the array setup and reformat?

File: packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 2]
```diff
@@ -173,24 +173,44 @@ private void handleVerifyPhoneNumber(
...
+                  .signInWithCredential(phoneAuthCredential)
+                  .addOnCompleteListener(
+                      new OnCompleteListener<AuthResult>() {
+
+                        @Override
+                        public void onComplete(@NonNull Task<AuthResult> task) {
+                          if (task.isSuccessful()) {
+                            Map<String, Object> arguments = new HashMap<>();
+                            arguments.put("handle", handle);
+                            channel.invokeMethod("phoneVerificationCompleted", arguments);
```
Comment (ened) [inline packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java]: Could the 2 onComplete handlers be merged?

File: packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 3]
```diff
@@ -173,24 +173,44 @@ private void handleVerifyPhoneNumber(
...
-                          Map<String, Object> arguments = new HashMap<>();
-                          arguments.put("handle", handle);
-                          channel.invokeMethod("phoneVerificationCompleted", arguments);
+            if (link) {
+              firebaseAuth
+                  .getCurrentUser()
+                  .linkWithCredential(phoneAuthCredential)
+                  .addOnCompleteListener(
+                      new OnCompleteListener<AuthResult>() {
+
```
Comment (ened) [inline packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java]: Please remove empty lines.

**Comment by cqhchan** [comment]
@ened  completed the requested changes

File: packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java (hunk 2 above)
Comment (cqhchan) [reply packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java]: Hi @ened  completed the requested changes

**Comment by cqhchan** [comment]
@tolotrasamuel 

`
  firebase_auth:
    git:
      url: git://github.com/cqhchan/plugins.git
      path: packages/firebase_auth
`

Insert this into pubspec.yaml to use it.  

https://github.com/cqhchan/plugins

This makes use of credentials implementation in https://github.com/flutter/plugins/pull/928 
I fixed a few bugs with their implementation and added a bool value in 
`
Future<void> verifyPhoneNumber(...)
`
to allow you to differentiate between verifyPhoneNumber for Login or For linking credentials with existing account. This addition bool value only affects specific android devices where they have auto verification.

**Comment by collinjackson** [comment]
Thanks for the contribution!

We've fixed the linkWithCredential issue (flutter/plugins#1333).

I'd love to get the other parts of your change, but I think we need an iOS implementation. Before we land this I'd ideally like to figure out how to write integration tests for phone auth so that we know the functionality is working on all platforms.

**Comment by collinjackson** [comment]
@cqhchan -- Please let me know if you'd like to implement the iOS side of the pull request or if you want to abandon this PR. Thanks!

**Comment by cqhchan** [comment]
hi @collinjackson , Im not too sure what IOS side u are referring to. The bug only appeared in Android. Regarding the additional feature to enable auto phone number verification, again it is an Android only feature in selected Android phones that support it

**Comment by creativecreatorormaybenot** [comment]
@collinjackson Is this going to be merged soon then? (considering the comment)

**Comment by collinjackson** [comment]
The API design in #1568 seems preferable to me, so I'm closing this one.

**Pull request closed without merging**